project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/settings/MinimizationPluginState.kt

```diff
@@ -5,35 +5,61 @@ import org.plan.research.minimization.plugin.model.FunctionLevelStage
 import org.plan.research.minimization.plugin.model.MinimizationStage
 import org.plan.research.minimization.plugin.model.state.*
 
-import com.intellij.openapi.components.BaseState
+import com.intellij.openapi.components.*
 import com.intellij.util.xmlb.annotations.Property
 import com.intellij.util.xmlb.annotations.XCollection
+import org.jetbrains.annotations.NonNls
 
-class MinimizationPluginState : BaseState() {
-    var currentCompilationStrategy by enum<CompilationStrategy>(CompilationStrategy.GRADLE_IDEA)
+@Service(Service.Level.PROJECT)
+@State(
+    name = "org.intellij.sdk.settings.AppSettings",
+    storages = [Storage("ProjectMinimizationSettings.xml")],
+)
+class MinimizationPluginState : PersistentStateComponent<MinimizationPluginState.State>, BaseState() {
+    private var myState = State()
+    var stateObservable: MinimizationPluginStateObservable = MinimizationPluginStateObservable(this)
 
-    /**
-     * A location for cloned projects
-     */
-    var temporaryProjectLocation by string("minimization-project-snapshots")
-    var snapshotStrategy by enum<SnapshotStrategy>(SnapshotStrategy.PROJECT_CLONING)
+    override fun getState(): State = myState
 
-    @Property(surroundWithTag = false)
-    @XCollection(style = XCollection.Style.v1, elementName = "stage")
-    var stages: MutableList<MinimizationStage> = mutableListOf(
-        FunctionLevelStage(
-            ddAlgorithm = DDStrategy.PROBABILISTIC_DD,
+    override fun loadState(state: State) {
+        myState = state
+    }
+
+    fun freezeSettings(isFrozen: Boolean) {
+        state.isFrozen = isFrozen
+    }
+
+    data class State(
+        @NonNls
+        var compilationStrategy: CompilationStrategy = CompilationStrategy.GRADLE_IDEA,
+        var temporaryProjectLocation: String = "minimization-project-snapshots",
+        var snapshotStrategy: SnapshotStrategy = SnapshotStrategy.PROJECT_CLONING,
+        var exceptionComparingStrategy: ExceptionComparingStrategy = ExceptionComparingStrategy.SIMPLE,
+
+        @Property(surroundWithTag = false)
+        @XCollection(style = XCollection.Style.v1, elementName = "stage")
+        var stages: MutableList<MinimizationStage> = mutableListOf(
```

Comment:
Let's make this immutable: `List<MinimizationStage>`. The problem for now: the changes of this list, such as `.add` or `.clear`, are not observable, so they won't affect observers. Making this list immutable (but `var` should be present, not `val`) solves this problem.